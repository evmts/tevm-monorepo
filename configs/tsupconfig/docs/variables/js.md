**@tevm/tsupconfig** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > js

# Variable: js

> **`const`** **js**: `Options`

## Source

[js.js:3](https://github.com/evmts/tevm-monorepo/blob/main/configs/tsupconfig/src/js.js#L3)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
