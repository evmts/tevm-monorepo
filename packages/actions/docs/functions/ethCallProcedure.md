[**@tevm/actions**](../README.md) • **Docs**

***

[@tevm/actions](../globals.md) / ethCallProcedure

# Function: ethCallProcedure()

> **ethCallProcedure**(`client`): [`EthCallJsonRpcProcedure`](../type-aliases/EthCallJsonRpcProcedure.md)

## Parameters

• **client**: `TevmNode`\<`"fork"` \| `"normal"`, `object`\>

## Returns

[`EthCallJsonRpcProcedure`](../type-aliases/EthCallJsonRpcProcedure.md)

## Defined in

[packages/actions/src/eth/ethCallProcedure.js:8](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/eth/ethCallProcedure.js#L8)
