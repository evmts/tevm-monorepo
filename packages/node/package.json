{
	"name": "@tevm/node",
	"version": "2.0.0-next.108",
	"private": false,
	"description": "Client implementations of Tevm",
	"keywords": [
		"solidity",
		"sol",
		"typescript",
		"web3",
		"blockchain"
	],
	"repository": {
		"type": "git",
		"url": "https://github.com/evmts/tevm-monorepo.git",
		"directory": "core/node"
	},
	"license": "MIT",
	"contributors": [
		"Will Cory <willcory10@gmail.com>"
	],
	"type": "module",
	"exports": {
		"./package.json": "./package.json",
		".": {
			"import": {
				"types": "./dist/index.d.ts",
				"default": "./dist/index.js"
			},
			"require": {
				"types": "./dist/index.d.cts",
				"default": "./dist/index.cjs"
			}
		}
	},
	"main": "dist/index.cjs",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"files": [
		"dist",
		"src",
		"!src/**/*.spec.ts",
		"types"
	],
	"scripts": {
		"all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs && bun lint:deps && bun lint:package",
		"build": "nx run-many --targets=build:dist,build:types --projects=@tevm/node",
		"build:dist": "tsup",
		"build:types": "tsup --dts-only",
		"clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
		"format": "biome format . --write",
		"format:check": "biome format .",
		"generate:docs": "typedoc",
		"lint": "biome check . --write --unsafe",
		"lint:check": "biome check . --verbose",
		"lint:deps": "bunx depcheck",
		"lint:package": "bunx publint --strict && attw --pack",
		"package:up": "pnpm up --latest",
		"test": "vitest --watch",
		"test:coverage": "vitest --coverage",
		"test:run": "vitest",
		"typecheck": "tsc --noEmit"
	},
	"dependencies": {
		"@tevm/address": "workspace:^",
		"@tevm/block": "workspace:^",
		"@tevm/blockchain": "workspace:^",
		"@tevm/common": "workspace:^",
		"@tevm/evm": "workspace:^",
		"@tevm/jsonrpc": "workspace:^",
		"@tevm/logger": "workspace:^",
		"@tevm/predeploys": "workspace:^",
		"@tevm/receipt-manager": "workspace:^",
		"@tevm/state": "workspace:^",
		"@tevm/sync-storage-persister": "workspace:^",
		"@tevm/tx": "workspace:^",
		"@tevm/txpool": "workspace:^",
		"@tevm/utils": "workspace:^",
		"@tevm/vm": "workspace:^"
	},
	"devDependencies": {
		"@tevm/test-utils": "workspace:^",
		"@tevm/tsconfig": "workspace:^",
		"@tevm/tsupconfig": "workspace:^"
	},
	"peerDependencies": {
		"viem": "^2.14.2"
	},
	"publishConfig": {
		"access": "public"
	},
	"sideEffect": false
}
