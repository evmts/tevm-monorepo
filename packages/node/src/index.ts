export { createTevmNode } from './createTevmNode.js'
export type { TevmNode } from './TevmNode.js'
export type { TevmNodeOptions } from './TevmNodeOptions.js'
export type { BaseClientOptions } from './BaseClientOptions.js'
export type { CustomPrecompile } from './CustomPrecompile.js'
export type { Extension } from './Extension.js'
export type { Hardfork } from './Hardfork.js'
export type {
	MiningConfig,
	AutoMining,
	ManualMining,
	IntervalMining,
} from './MiningConfig.js'
export { prefundedAccounts, GENESIS_STATE } from './GENESIS_STATE.js'
export type { Filter, FilterType } from './Filter.js'
export {
	ProviderRpcError,
	type EIP1193Events,
	type EIP1193EventMap,
	type ProviderMessage,
	type EIP1193EventEmitter,
	type ProviderConnectInfo,
} from './EIP1193EventEmitterTypes.js'
