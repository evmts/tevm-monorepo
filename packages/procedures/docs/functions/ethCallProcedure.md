[**@tevm/procedures**](../README.md) • **Docs**

***

[@tevm/procedures](../globals.md) / ethCallProcedure

# Function: ethCallProcedure()

> **ethCallProcedure**(`client`): [`EthCallJsonRpcProcedure`](../type-aliases/EthCallJsonRpcProcedure.md)

## Parameters

• **client**: `TevmNode`\<`"fork"` \| `"normal"`, `object`\>

## Returns

[`EthCallJsonRpcProcedure`](../type-aliases/EthCallJsonRpcProcedure.md)

## Defined in

[procedures/src/eth/ethCallProcedure.js:8](https://github.com/evmts/tevm-monorepo/blob/main/packages/procedures/src/eth/ethCallProcedure.js#L8)
