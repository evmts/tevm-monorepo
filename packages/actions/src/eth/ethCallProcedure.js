import { callProcedure } from '../call/callProcedure.js'

/**
 * Executes a message call without creating a transaction on the block chain.
 * @param {import('@tevm/node').TevmNode} client
 * @returns {import('./EthProcedure.js').EthCallJsonRpcProcedure}
 */
export const ethCallProcedure = (client) => async (req) => {
	const [tx, blockTag, stateOverrideSet, blockOverrideSet] = req.params
	const { data, from, to, gas, gasPrice, value } = tx
	const response = await callProcedure(client)({
		...(req.id !== undefined ? { id: req.id } : {}),
		jsonrpc: req.jsonrpc,
		method: 'tevm_call',
		params: [
			{
				...(gasPrice !== undefined ? { gasPrice } : {}),
				...(data !== undefined ? { data } : {}),
				...(gas !== undefined ? { gas } : {}),
				...(value !== undefined ? { value } : {}),
				...(to !== undefined ? { to } : {}),
				...(from !== undefined ? { from } : {}),
				...(blockTag !== undefined ? { blockTag } : {}),
			},
			stateOverrideSet,
			blockOverrideSet,
		],
	})
	if (!response.result) {
		return {
			jsonrpc: req.jsonrpc,
			method: 'eth_call',
			error: response.error,
			...(req.id !== undefined ? { id: req.id } : {}),
		}
	}
	return {
		jsonrpc: req.jsonrpc,
		method: 'eth_call',
		result: response.result.rawData,
		...(req.id !== undefined ? { id: req.id } : {}),
	}
}
