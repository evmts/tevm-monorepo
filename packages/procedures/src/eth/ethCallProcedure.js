import { callProcedure } from '../index.js'

/**
 * Executes a message call without creating a transaction on the block chain.
 * @param {import('@tevm/base-client').BaseClient} client
 * @returns {import('@tevm/procedures-types').EthCallJsonRpcProcedure}
 */
export const ethCallProcedure = (client) => async (req) => {
	const [tx, blockTag] = req.params
	const { data, from, to, gas, gasPrice, value } = tx
	const response = await callProcedure(client)({
		...(req.id !== undefined ? { id: req.id } : {}),
		jsonrpc: req.jsonrpc,
		method: 'tevm_call',
		params: [
			{
				...(gasPrice !== undefined ? { gasPrice } : {}),
				...(data !== undefined ? { data } : {}),
				...(gas !== undefined ? { gas } : {}),
				...(value !== undefined ? { value } : {}),
				...(to !== undefined ? { to } : {}),
				...(from !== undefined ? { from } : {}),
				...(blockTag !== undefined ? { blockTag } : {}),
			},
		],
	})
	if (!response.result) {
		return {
			jsonrpc: req.jsonrpc,
			method: 'eth_call',
			error: response.error,
			...(req.id !== undefined ? { id: req.id } : {}),
		}
	}
	return {
		jsonrpc: req.jsonrpc,
		method: 'eth_call',
		result: response.result.rawData,
		...(req.id !== undefined ? { id: req.id } : {}),
	}
}
