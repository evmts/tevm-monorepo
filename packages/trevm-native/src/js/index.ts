import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};