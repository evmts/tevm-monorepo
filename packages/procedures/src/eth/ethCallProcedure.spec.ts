import { type TevmNode, createTevmNode } from '@tevm/node'
import { beforeEach, describe, expect, it } from 'vitest'
import type { EthCallJsonRpcRequest } from './EthJsonRpcRequest.js'
import { ethCallProcedure } from './ethCallProcedure.js'

let client: TevmNode

beforeEach(() => {
	client = createTevmNode()
})

describe('ethCallProcedure', () => {
	it('should execute a message call successfully', async () => {
		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			id: 1,
			params: [
				{
					data: '0x0',
					from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
					to: `0x${'69'.repeat(20)}`,
					value: '0x1',
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_call')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle missing optional fields in the request', async () => {
		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			id: 1,
			params: [
				{
					from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
					to: `0x${'69'.repeat(20)}`,
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_call')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle requests without an id', async () => {
		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			params: [
				{
					from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
					to: `0x${'69'.repeat(20)}`,
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_call')
		expect(response.id).toBeUndefined()
		expect(response.result).toMatchSnapshot()
	})
})
