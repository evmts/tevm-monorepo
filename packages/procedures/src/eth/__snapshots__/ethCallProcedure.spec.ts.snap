// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ethCallProcedure should return error when call fails 1`] = `
{
  "code": -32000,
  "data": {
    "errors": [
      
"sender doesn't have enough funds to send tx. The upfront cost is: 210000000 and the sender's account (0x6969696969696969696969696969696969696969) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0xd3e22d4c76284ef1cd4b984bc3e20308dbcb13a96239f9e550ca74b03d67f779 nonce=0 value=0 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
    ],
  },
  "message": 
"sender doesn't have enough funds to send tx. The upfront cost is: 210000000 and the sender's account (0x6969696969696969696969696969696969696969) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0xd3e22d4c76284ef1cd4b984bc3e20308dbcb13a96239f9e550ca74b03d67f779 nonce=0 value=0 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
}
`;

exports[`ethCallProcedure should handle missing optional fields in the request 1`] = `"0x"`;

exports[`ethCallProcedure should handle requests without an id 1`] = `"0x"`;

exports[`ethCallProcedure should execute a message call successfully 1`] = `"0x"`;
